File: src/main/scala/2017/day-8.scala
package adventofcode2017

import scala.io.Source

object Day8 {
  private val inputRowRegex = """([a-z]+) ([a-z]+) (-?[0-9]+) if ([a-z]+) (.+) (-?[0-9]+)""".r
  lazy val input: String = Source.fromFile(s"2017/input/day-8.input").mkString.trim
  val equalities = Map[String, (Int, Int) => Boolean](
    ">" -> (_>_),
    "<" -> (_<_),
    ">=" -> (_>=_),
    "<=" -> (_<=_),
    "==" -> (_==_),
    "!=" -> (_!=_)
  )
  val oprs = Map[String, (Int, Int) => Int](
    "inc" -> (_+_),
    "dec" -> (_-_)
  )
  def main(args: Array[String]): Unit = {
    val (map, max) = input.lines.foldLeft((Map[String, Int](), 0)) {
      case ((map, max), inputRowRegex(reg, opr, amount, test, op, value)) => 
        if (equalities(op)(map.getOrElse(test, 0), value.toInt)) {
          val newVal = oprs(opr)(map.getOrElse(reg, 0), amount.toInt)
          (map + (reg -> newVal), Math.max(max, newVal))
        } else {
          (map, max)
        }
    }
    println(map.values.max)
    println(max)
  }
}

File: 2017/js/day-8.js
const fs = require('fs');
const input = fs.readFileSync('2017/input/day-8.input').toString();

const instructions = input.split('\n');
let map = {};
const equalities = {
  '>': (a, b) => a > b,
  '>=': (a, b) => a >= b,
  '<': (a, b) => a < b,
  '<=': (a, b) => a <= b,
  '!=': (a, b) => a != b,
  '==': (a, b) => a == b,
};

const oprs = {
  'inc': (a, v) => a + v,
  'dec': (a, v) => a - v,
};

let max = 0;
instructions.forEach(entry => {
  const parts = entry.split(' ');
  if (equalities[parts[5]](map[parts[4]] || 0, parts[6])) {
    const value = map[parts[0]] || 0;
    max = Math.max(max, value);
    map[parts[0]] = oprs[parts[1]](value, +parts[2]);
  }
});

const result = Math.max(...Object.keys(map).map(key => map[key]));
console.log(result);
console.log(max);
